File: vuln-sandbox/python/cwe-89/sql-injection-safe.py
# SAFE: sql-injection — parameterized query prevents SQL injection
# Rule: SqlStringConcat | CWE-89 | Expected: TrueNegative

import sqlite3
from flask import Flask, request, jsonify

app = Flask(__name__)
DB_PATH = 'users.db'


def get_db():
    return sqlite3.connect(DB_PATH)


@app.route('/api/users/<user_id>', methods=['GET'])
def get_user(user_id):
    """Fetch a user by ID using a parameterized query."""
    conn = get_db()
    cursor = conn.cursor()

    # SAFE: user_id passed as a parameter placeholder; never concatenated into the query string
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))

    row = cursor.fetchone()
    conn.close()

    if row is None:
        return jsonify({'error': 'User not found'}), 404
    return jsonify({'id': row[0], 'username': row[1], 'email': row[2]})


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/typescript/cwe-89/sql-injection-safe.ts
// SAFE: sql-injection — parameterized query prevents SQL injection (TypeScript)
// Rule: SqlStringConcat | CWE-89 | Expected: TrueNegative

import express, { Request, Response } from 'express';
import mysql2 from 'mysql2/promise';

const app = express();

async function getUser(req: Request, res: Response): Promise<void> {
  const db = await mysql2.createConnection({
    host: 'localhost',
    user: 'root',
    database: 'app',
  });

  const userId: string = req.params.id;

  // SAFE: parameterized placeholder used; user input never concatenated into the query
  const [rows] = await db.query('SELECT * FROM users WHERE id = ?', [userId]);

  await db.end();
  res.json(rows);
}

app.get('/user/:id', getUser);
app.listen(3000);
